### PR TITLE
Feature | Web push notifications setup

### PR DESCRIPTION
**Pull request by AlanSoto31**
## Description
- Setup web push notifications for Session reminders using the gem [web-push](https://github.com/pushpad/web-push).
- Add  the `webpush` delivery method to make web push notifications to work with Noticed gem.
- The current setup supports sending web push notifications on Chrome and Firefox for both mobile and desktop devices, and on Safari for desktop only. A PWA is required for iPhone compatibility. More details can be found [here](https://webkit.org/blog/13878/web-push-for-web-apps-on-ios-and-ipados/) and [here](https://developer.apple.com/documentation/usernotifications/sending-web-push-notifications-in-web-apps-and-browsers).

NOTES:
- Sync the `service-worker.js` files on PR https://github.com/TelosLabs/rails-world/pull/76 to make the notifications work in the installed app.
- Look at [this diagram](https://excalidraw.com/#json=1KzYqGw7ZdRoHIGAImyVG,Bx60pTgWWbTkxPmV13DmfA) of the stimulus controller that handles the permissions and subscription logic.
- This [post](https://web.dev/articles/push-notifications-overview) helped me to understand the flow of web push notifications.

## How has this been tested?

Please mark the tests that you ran to verify your changes. If difficult to test, consider providing instructions so reviewers can test.

- [x] Manual testing
- [ ] System tests
- [ ] Unit tests
- [ ] None

## Checklist

- [x] CI pipeline is passing
- [ ] My code follows the conventions of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (if applicable)
- [ ] I have added seed data to the database (if applicable)

## Release tasks

Add any tasks that need to be done before/after the release of this feature.

### Desktop
image

### IOS
WhatsApp Image 2024-08-06 at 15 03 14

### Android
WhatsApp Image 2024-08-06 at 15 04 21

### CODE REVIEW

File: config/credentials.yml.enc [hunk 1]
```diff
@@ -1 +1 @@
-BZ7hJ1Q/E6O5OSrGEVEqVzH7xhDQ6/unNX6tY6eYmTCo+K8kymS70b/1fHFaZiYLbWqx4tc/73L0irvjg6W0wB5B+a8SCgvjo2wP1VqfFEeBNlPJI82ZJdBFkz30dsBgf1gw6Rraj3Ze4e/IcMJQo8UP8I5x+Ps2IPrEYGlhUHi8cBR8R11qIxHHattpfxqQ7M89C8l+hsjf793ZyLnt9TmFZdxWbxnBDdZHhYSpmJTVV34mFDP9qZ7hPYTGXzGqhpszIP+6l59pmTpyBO/Pq21eX7MpQi+XwZCCxTGAq5ih0TgiBwf9IAhiTgZ7/96vVYrcU+7h4KMnJk19hwl3NHXuaWo6uET3F8Munc/qTz/ShXxCX3zX2g57fyPkf6pzXA6QXGzf8flemmesGUjYJbKZX1VZ--tlJOD/9KN6MF1juj--IdnvUnXcOke5Wp5YlKghwg==
\ No newline at end of file
+JfIGc8QNiYbht8i2O0bM2MQ5gGdbihW9nFMYmOn5AovjrsByy7SnD3bI1jJCKlnAZJrCpmXNFXsaLnOnFrVwF+u5KeE2K9E8GC1LGgeAErCagmS18MOgKldj+IjnR9qIZPkQVr2LSwC286/bAsM+Jh6CO3jFamjIQ3rZc5zQHHIITuLz8EUfgqXAEDHmn7Q+MiGc8igGn1PWPKI14MCE2FJGb/UvGn75pmTkRv1Gu2KmDhJ8II4MIjgXoVntATgQ28SnhWdEDVMJolzNknn2MZ744kVzkqtRDfAkYJpL0AgOoFSSQXqkUEWItWJWsTz/na5FLh3oRzNhwl3F3Oc48rhvMrU7jIIzd2oVdpwy1+JCaxeCDEfy1PF1VsIU9OEnu8pr7Aal2zcnbk0bPdw4Ap1o5PMUpW1p+oXhyXD9+ChNU1YpFgSJNojqKesBfCPIUCmMFo6urLHiEHabfWm/TNWOt4dVfglBUvuP6mccjx45MAw/9SoXrXsa+yMDdWPXZ+rYXVfyCDXArsyOcxI5ZmOKAZ8mgrAGJynrk8944EJdeTpw08AnVCkKFFO/cjbB5w1hVAG2uU9D1WrCy257m+rKDiO4m9gLSNQT/X/all68eYlXDCmlCuUpqf98qYgODg==--/o6oYJBdrlpXzX07--/z6lx4lf08Maf88vlPsaYQ==
```
Comment (AlanSoto31) [inline config/credentials.yml.enc:1]: Adding [VAPID](https://pushpad.xyz/blog/web-push-what-is-vapid)

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -5,8 +5,16 @@ class ApplicationController < ActionController::Base
...
   private
 
   def current_profile = current_user&.profile
+
+  def vapid_public_key
+    return nil if Rails.env.test?
+
+    vapid_key = Rails.application.credentials.dig(:vapid, :public_key)
+    Base64.urlsafe_decode64(vapid_key).bytes
+  end
```
Comment (AlanSoto31) [inline app/controllers/application_controller.rb:19]: Decoding the vapid key in test environment causes the system specs to fail.

File: app/javascript/controllers/web_push_notifications_controller.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+      console.error('Service workers are not supported by this browser.')
+      return
+    }
+
+    if (!this.vapidKeyValue) {
+      console.error('VAPID key is not available.')
+      return
+    }
+
+    await navigator.serviceWorker.register('/service_worker.js')
```
Comment (Sergio-e) [inline app/javascript/controllers/web_push_notifications_controller.js:75]: We should register a service worker only if one doesn't exist already

File: app/controllers/web_push_subscriptions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+      p256dh: params.dig(:keys, :p256dh),
+      auth: params.dig(:keys, :auth),
+      user_id: current_user.id
+    )
+
+    if web_push_subscription.save
+      render json: web_push_subscription, status: :created
+    else
+      render json: web_push_subscription.errors, status: :unprocessable_entity
+    end
```
Comment (Sergio-e) [inline app/controllers/web_push_subscriptions_controller.rb:14]: The client isn't handling the success/failure response. I think it's ok if we go with `save!` for the moment

File: config/credentials.yml.enc [hunk 5]
```diff
@@ -1 +1 @@
-irbDWudk4jCs1xkxpK/cNdd9UHkVP+Sk8XlMLXiFl4636ZMZQlGRIJLRcK7OS12gjeCDkS6Ac9/z0IwTyzT1yYp0Hdd8bDgwDS/0gK/2/kGjfAfz1fRaYc+KbMRMIERxpgdL4wceMxAaW2XvpM+POUzPIaJDFj5l5t1iLoEYbBhPI6AHLyxhivSKnmixowiA4gerDgTkXQ==--+44A5a8oAobbmQDx--E4wpoVDDrY0JE13R3AV2Gw==
\ No newline at end of file
+cdtdCBDkc4VAQyYXppyxi/A+j9CrIIPzah3M2LFzsVTrABfZspckGL32cCekAQPL4tZqgikzX01IeAjzMcPkB1HXEb5sXqFZs+ZfXfCw46Xq34yNp/SFV43oV91Pn7M8nLmH6Dohx3N+jgkm1yWd5+0wS2e6JuguDA02MZtIhjhLuVrKhTLP/BEZeTLhKsyp9hLpbAjNDw==--qaPwxiYyEhDkNeOG--k8F4XSY7+pbfMa9dL6erfA==
```
Comment (Sergio-e) [inline config/credentials.yml.enc]: Reminder - remove file

File: app/javascript/controllers/web_push_notifications_controller.js [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  handleDeniedPermission () {
+    if (!this.inAppNotificationsValue) return
+
+    this.displayNotificationBlockingMessage()
+  }
+
+  handleDefaultPermission () {
+    if (this.hasEnableNotificationsTarget) {
```
Comment (Sergio-e) [inline app/javascript/controllers/web_push_notifications_controller.js:30]: Let's go with two `if` statements

**Review by Sergio-e — COMMENTED** [review]
💪

**Review by Sergio-e — APPROVED** [review]

**Pull request merged**